pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+            # so better to raise an error
+            raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+
+        return DmapScalar(scalar_name, scalar_value, scalar_type)
+
+    def read_array(self, record_size) -> DmapArray:
+        """
+        Reads an array from a dmap record the byte arrays and
+        stores the data properties in a DmapArray structure.
+
+        Return
+        ------
+        DmapArray : namedtuple
+             data structure that contains the data properties of
+             the array read in.
+
+        Raises
+        -------
+            DmapDataError
+                if the array properties (like a dimension size == 0)
+                are incorrect.
+
+        See Also
+        --------
+        read_data : reads the data in the byte array
+        read_numerical_array : reads in a numerical array from the byte array
+        read_string_array: reads in a string/DMAP array
+        """
+        array_name = self.read_data('s', 1)
+        array_type = self.read_data('c', 1)
+
+        self.check_data_type(array_type, array_name)
+
+        array_type_fmt = DMAP_DATA_TYPES[array_type][0]
+        array_fmt_bytes = DMAP_DATA_TYPES[array_type][1]
+
+        array_dimension = self.read_data('i', 4)
+        self.bytes_check(array_dimension, "array dimension",
+                         record_size, "record size")
+        self.zero_check(array_dimension, "array dimension")
+
+        array_shape = [self.read_data('i', 4) for i in range(0, array_dimension)]
+        array_shape.reverse()
+
+        # if shape list is empty
+        if len(array_shape) != array_dimension:
+            message = "Error: Array shape {shape} could not be read."\
+                    " cursor: {cursor}".format(shape=array_shape,
+                                               cursor=self.cursor)
+            raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+        # TODO Explain where does slist come from?
+        elif any(x <= 0 for x in array_shape) and array_name != "slist":
+            message = "Error: Array shape {shape} contains "\
+                    "dimension size <= 0."\
+                    " Cursor: {cursor}".format(dim=array_shape,
+                                               cursor=self.cursor)
+            raise pydmap_exceptions.DmapDataError(self.filename, message)
+
+        for i in range(array_dimension):
+            if array_shape[i] >= record_size:
+                message = "Error: Array {index}-dimension size {size}"\
+                        " exceeds record size: {rec_size}."\
+                        "".format(index=i,
+                                  size=array_shape[i],
+                                  rec_size=record_size)
+                raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+
+        # We could use np.prod(array_shape) but the for loop has a better
+        # time performance.
+        total_num_cells = 1
```

Comment:
Maybe I wrote this, but what is meant by cells? maybe a more clear name would be better here.

Reply:
You had `total_elements` and I felt elements from a mathematical standpoint was inaccurate. So I looked up matrix terms, and the number cells are equivalent to what you called the number of elements. Plus, I used elements earlier on as something else, so I didn't want break consistency.  

I left a comment to make it more clear.

Reply:
Sounds good. total_matrix_cells would work good as a name too.